User can register for an account CLIENT

ViktorHek opened the pull request:
[User can register for an account CLIENT](https://www.pivotaltracker.com/story/show/176470274)
### User Story
```
As an User,
In order to subscribe,
We would like to authenticate myself.
```
## Changes proposed in this pull request:
* Creating a test for visitors can register to read articles
* Creating RegistrationForm using Modal
* Adding logic to hide Sign Up button when registration was successful
* Implementing a function to make registration form displayed when a visitor clicks an article

## Packages/Gems added
* j-tockauth

## Screenshots (Client)
<img width="1440" alt="Screenshot 2021-01-13 at 20 22 11" src="https://user-images.githubusercontent.com/72070417/104499207-0a11bc80-55dd-11eb-894d-9e9dcff804fe.png">

ViktorHek committed:
creating user can register test. creating user can register fixture. creating RegistrationForm. creating authentication module

ViktorHek committed:
changing error/success message. adding case SET_CURRENT_USER

salindholm committed:
Importing useDispatch and signup module to registrationform and initializing jtockauth

salindholm committed:
Importing jtockauth, creating a post request in authentication module with catch error handler

salindholm committed:
Adding index article stub to test, updating password confirmation input correctly and adding password confirmation event target value to authentication module

salindholm committed:
Merge branch 'development' of https://github.com/CraftAcademy/kitty_news_client_user into user_can_register_to_buy_subscription

salindholm committed:
Updating data cy and test, adding Modal with button and message. Imporint useSelector, useState and implementing it.

salindholm committed:
Adding sad path stub in test, removing registrationForm component from app.jsx to categoryMenu component. Changing currentUser state to errorMessage, changing name on action type in module and removing success message action.payload from rootreducer

KyunginNa committed:
Modifying userCanRegisterToBuySubscription.feature.js
Formatting code
Removing debugger

KyunginNa committed:
Adding setOpen in onClick event in Registration.jsx
[WIP]Trying to fix mocked resopnse in test because sad path is failing

ViktorHek commented on cypress/integration/userCanRegisterToBuySubscription.feature.js:
```diff
@@ -0,0 +1,62 @@
+      });
+      cy.visit("/");
+    });
+    it("with invalid credentials", () => {
```
add a sad path for "this email is already taken"

KyunginNa replied:
We thought it'd be okay not to test all the sad paths but just one, because it is built-in function in rails and we all know it works well...

KyunginNa replied:
I'll add it! :)

ViktorHek commented on cypress/integration/userCanRegisterToBuySubscription.feature.js:
```diff
@@ -0,0 +1,62 @@
+      cy.visit("/");
+    });
+    it("with valid credentials", () => {
+      cy.get("[data-cy='signup-button']").click();
+      cy.get("[data-cy='input-email']").type("registered@user.com");
+      cy.get("[data-cy='input-password']").type("password");
+      cy.get("[data-cy='input-password-confirmation']").type("password");
+      cy.get("[data-cy='submit-btn']").click();
```
ad this inside ```cy.get("[data-cy='registration-form']").within(() => {}```

KyunginNa replied:
Added!

KyunginNa committed:
Fixing mocked response in test

KyunginNa committed:
Changing color of button

ViktorHek committed:
modify readSpecificArticle test so test loges in. adding sand path in userCanRegisterToReadArticles. adding currentUser to useSelector and useEffect in displayArticle. adding another error message called registerErrorMessage. adding action in rootReducer called REGISTER_ERROR_MESSAGE. adding if currentUser in show action

psnoya committed:
Modyfied userCanReadASpecificArticle and userCanViewArticlesByCategory teste

KyunginNa committed:
Modifying userCanReadASpecificArticle test
Implementing a function to make registration form popped up when a visitor clicks an article
Adding OPEN_MODAL, CLOSE_MODAL action in rootReducer.js

KyunginNa committed:
Adding logic to hide sign up button when current user exists
and adding a test for it

KyunginNa committed:
Adding currentUser in useEffect dependency

ViktorHek commented on src/components/RegistrationForm.jsx:
```diff
@@ -0,0 +1,74 @@
+    <>
+      <Modal
+        closeIcon
+        onClose={() => dispatch({ type: "CLOSE_MODAL" })}
+        onOpen={() => dispatch({ type: "OPEN_MODAL" })}
```
not really a problem a.t.m but when we add the `subscriber modal` we might want to change this name "CLOSE_REGISTATION" and "OPEN_REGISTRATION"

KyunginNa replied:
Good thinking!⭐️ Changed them to OPEN_REGISTRATION_FORM and CLOSE_REGISTRATION_FORM

KyunginNa committed:
Changing action names: OPEN_MODAL to OPEN_REGISTRATION_FORM, CLOSE_MODAL to CLOSE_REGISTRATION_FORM

emtalen commented on cypress/integration/userCanReadASpecificArticle.feature.js:
```diff
@@ -37,7 +56,7 @@ describe("User can", () => {
   });
 
-  context("not see an article if improper params are provided", () => {
+  describe("not see an article if improper params are provided", () => {
```
why did you change this from `context` to `describe`?

KyunginNa replied:
We have `describe` in other tests, so it was just for consistency!

emtalen commented on cypress/integration/userCanReadASpecificArticle.feature.js:
```diff
@@ -58,4 +94,12 @@ describe("User can", () => {
     });
   });
+
+  describe("NOT read article when they're not registered", () => {
```
Make sure that this is restricted in your API as well. So not anything you need to change in this PR but take a note on updating your API

emtalen commented on src/modules/getArticles.js:
```diff
@@ -13,12 +13,19 @@ const getArticles = {
+        });
+      }
+    } else {
+      dispatch({ type: "OPEN_REGISTRATION_FORM" });
```
nice solution for this ✨
